File: ACM/2010/Final/C - Cocircular Points/cocircular.cpp
#include <cstdio>
#include <cmath>
#include <map>
#include <set>
#include<vector>
#include<algorithm>


#define EPS 1e-6

using namespace std;

inline int cmp(double x, double y) {
   return (x <= y + EPS) ? (x + EPS < y) ? -1 : 0 : 1;
}

struct point {
   double x, y;
   point(double x = 0, double y = 0): x(x), y(y) {}
   point operator +(const point &q) { return point(x + q.x, y + q.y); }
   point operator -(const point &q) { return point(x - q.x, y - q.y); }
   point operator *(const double &t) { return point(x * t, y * t); }
   point operator /(const double &t) { return point(x / t, y / t); }
   double operator *(const point &q) { return x * q.x + y * q.y; }
   double operator %(const point &q) { return x * q.y - y * q.x; }
   int cmp(const point &q) const {
   if (int t = ::cmp(x, q.x)) return t;
   return ::cmp(y, q.y);
   }
   bool operator ==(const point &q) const { return cmp(q) == 0; }
   bool operator !=(const point &q) const { return cmp(q) != 0; }
   bool operator < (const point &q) const { return cmp(q) < 0; }

};

point circumcenter(point &p, point &q, point &r) {
   point a = p - r, b = q - r, c = point(a * (p + r) / 2, b * (q + r) / 2);
   return point(c % point(a.y, b.y), point(a.x, b.x) % c) / (a % b);
}

double norma(const point &p) { return hypot(p.x, p.y); }

point points[105];

int i, j;


int main(){
	int n,x,y;
	while(true){       
		scanf("%d",&n);
		if(!n)
			break;
		for(int i=0; i<n; i++){;
			scanf("%lf",&points[i].x);
			scanf("%lf",&points[i].y);
		}
		if(n<3)
			printf("%d\n",n);
		int maxc = 0;
		vector<point> centers;
		for(int i=0; i<n; i++) {
			for(int j=i+1; j<n; j++) {
				/*for(int k=j+1; k<n; k++){
					point A=points[j]-points[i], B=points[j]-points[k];
					if(cmp(abs(A*B),abs(norma(A)*norma(B)))==0){
						continue;
					}
					point center=circumcenter(points[i],points[j],points[k]);
					centers.push_back(center);
				}*/
				sort(points.begin(), pon.end());
				/*for (int k = 1; k < centers.size(); k++) {
					printf("(%.2lf %.2lf)\n", centers[k].x, centers[k].y);
				}
				printf("----------------\n");
				*/int conta = 1;
				for (int k = 1; k < centers.size(); k++) {
					if (centers[k] == centers[k-1])
						conta++;
					else {
						if (conta > maxc)
							maxc = conta;					
						conta = 1;
					} 
				}
				
				                
			}
		}
		printf("%d\n",maxc+2);
	}
}

File: ACM/2010/Final/C - Cocircular Points/cocircular.java

import java.util.*;
import java.io.*;

public class cocircular {

    static int N;
    static Point[] pontos = new Point[110];

    public static void main(String args[]) throws Exception {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        while (true) {
            String[] line = in.readLine().split("\\s");
            N = Integer.parseInt(line[0]);
            if (N == 0) {
                break;
            }
            for (int i = 0; i < N; i++) {
                line = in.readLine().split("\\s");
                int x = Integer.parseInt(line[0]);
                int y = Integer.parseInt(line[1]);
                pontos[i] = new Point(x, y);
            }
            if (N < 3) {
                System.out.printf("%d\n", N);
                continue;
            }
			Map<Circle, Set<Integer> > counter = new HashMap<Circle, Set<Integer>>();
            int maxc = 0;
            for (int i = 0; i < N; i++) {
                for (int j = i + 1; j < N; j++) {
                    for (int k = j + 1; k < N; k++) {    
						Point A = pontos[j].sub(pontos[i]), B = pontos[j].sub(pontos[k]);
						if (cmp(Math.abs(A.dot(B)), Math.abs(A.norma() * B.norma())) == 0) {
                           continue;
                        }
                        Circle c = Circle.build(pontos[i], pontos[j], pontos[k]);
						Set<Integer> ss = null;
                        if ((ss = counter.get(c)) == null) {
							ss = new HashSet<Integer>();
							counter.put(c, ss);
						}
						//ss.add(i);
						//ss.add(j);
						//ss.add(k);
						//if (ss.size() > maxc)
						//	maxc = ss.size();
                    }
                }
            }
            System.out.println(maxc < 2 ? 2 : maxc);
        }
    }

    static int cmp(double x, double y) {
       return (x <= y + EPS) ? (x + EPS < y) ? -1 : 0 : 1;
    }
    static final double EPS = 1e-6;

    static class Point {

        double x, y;

        public Point(double x, double y) {
            this.x = x;
            this.y = y;
        }

        public Point sub(Point p) {
            return new Point(x - p.x, y - p.y);
        }

        public Point add(Point p) {
            return new Point(x + p.x, y + p.y);
        }

        public Point mult(double d) {
            return new Point(x * d, y * d);
        }

        public Point div(double d) {
            return new Point(x / d, y / d);
        }

        public double dot(Point p) {
            return (x * p.x + y * p.y);
        }

        public double cross(Point p) {
            return (x * p.y - y * p.x);
        }

        public double norma() {
            return Math.sqrt(x*x+y*y);
        }

        public double dist(Point p) {
            return (this.sub(p).norma());
        }

        public String toString() {
            return String.format("(%.4f %.4f)", x, y);
        }

        @Override
        public boolean equals(Object obj) {
            if (obj == null) {
                return false;
            }
            final Point other = (Point) obj;
            if (cmp(this.x, other.x) == 0 && cmp(this.y, other.y) == 0) {
                return true;
            }
            return false;
        }

        @Override
        public int hashCode() {
            int hash = 5;
            hash = 29 * hash + ((((int) x) ^ ((int) x)) >>> 3);
            hash = 29 * hash + ((((int) x) ^ ((int) x)) >>> 3);
            return hash;
        }
    }

    static class Circle implements Comparable<Circle>{
        Point center;
        double radius;

        public Circle(Point c, double r) {
            center = c;
            radius = r;
        }

		public int compareTo(Circle other) {
			return -1;
		} 

        static Circle build(Point p, Point q, Point r) {
            Point a = p.sub(r), b = q.sub(r), c = new Point(a.dot(p.add(r)) / 2, b.dot(q.add(r)) / 2);
            Point center = new Point(c.cross(new Point(a.y, b.y)), new Point(a.x, b.x).cross(c)).div(a.cross(b));
            return new Circle(center, center.dist(p));
        }

        @Override
        public boolean equals(Object obj) {
            if (obj == null) {
                return false;
            }
            final Circle other = (Circle) obj;
            if (cmp(this.radius, other.radius) == 0 && this.center.equals(other.center)) {
                return true;
            }
            return false;
        }

        @Override
        public int hashCode() {
            return (int) this.radius;
        }
    }
}
